### ISSUE TITLE
Change curl to wget

### ISSUE THREAD

**Issue by elijahgagne** [mattieb/download-windows-esd#3]
I ran into an issue where the download-windows-esd script uses curl to download the file. I'd get to 100%, but it would then error with

```
curl: (18) transfer closed with 26485 bytes remaining to read
```

I changed the script to use wget and that issue was resolved. Here's my full notes on what worked for me.

```
brew install wimlib

cd ~/Downloads
git clone https://github.com/mattieb/download-windows-esd.git
git clone https://github.com/mattieb/windows-esd-to-iso.git

sed -i '' 's@curl -#OR -C - "${url}"@wget "${url}"@g' ./download-windows-esd/download-windows-esd

./download-windows-esd/download-windows-esd download en-us Enterprise ARM64

./windows-esd-to-iso/windows-esd-to-iso $(ls *.esd)
```

Maybe just an issue on my end so you can close this, but I figured I'd post it in case it helps someone else.

Thanks!
-EWG

**Comment by mattieb** [comment]
Even if it is just on your end, I was curious about it. 🙂 

I have not done a full download in awhile since the images haven't updated, though every so often I do run the script in my directory with already-downloaded files; the resume will hit 100% immediately if they already exist, then checksum.

So I tried a brand-new download of one of the en-us images, and still didn't have the problem.

Curious what your download parameters are—i.e. language, edition, architecture?

**Comment by elijahgagne** [comment]
Hmm, I'm not immediately sure how I would tell what HTTP headers `curl` is using. I guess I could do a tcpdump. Let me know if you happen to know an easy way to figure that out. 

Here's me reproducing it outside of your code:

```sh
[d92495j:Downloads] % curl -#OR -C - "http://dl.delivery.mp.microsoft.com/filestreamingservice/files/3586ef88-7144-42e3-ba81-bfa35cd48762/22631.2861.231204-0538.23H2_NI_RELEASE_SVC_REFRESH_CLIENTBUSINESS_VOL_A64FRE_en-us.esd"
#########################################################################################################################################################  100.0%curl: (18) transfer closed with 3317 bytes remaining to read
[d92495j:Downloads] 4m1s 18 %
```

**Comment by mattieb** [comment]
Ah, no, I was asking about the parameters supplied to the script. But I can tell from the URL. Mostly I was curious if you were downloading an image that I had not, which isn't the case. 

I can run that same command line without issue; I'm inclined to think it's environmental, but also, I think a nice thing to do would be to add another couple subcommands:

1. "url", which, instead of downloading, will output a download URL, and
2. "shasum", which would output the checksum in a "shasum -c"-compatible-manner (very important especially since the downloads run over plain HTTP!)

I don't know if I'll do these anytime soon, but I'd like to.

**Comment by elijahgagne** [comment]
Yeah those would be nice to haves. In regards to #1, I don't have strong feelings here, but I will say that I do like how `wget` shows the URL in case you need it. I was trying to make the argument for you to switch from `curl` to `wget` and I came across https://stackoverflow.com/questions/636339/what-is-better-curl-or-wget where one of the commenters posted about how `curl` will sometimes result in `error 18 - transfer closed`, whereas `wget` does not. 

For #2, I can't believe that Microsoft is hosted that over HTTP in 2024! Wow

**Comment by mattieb** [comment]
Re: using wget, I don't have anything against it—but it's not included in macOS, so making it the default means everyone needs to take an extra step to get going.

That SO post you found is interesting. It makes me wonder if a curl flag might improve the situation for you. I can't reproduce it on any network I have access to, so anything I try would sadly be a shot in the dark, but if you can figure out a curl configuration that resolves your problem, I'm happy to evaluate it for inclusion.

**Comment by mattieb** [comment]
Added the "shasum" and "url" options in https://github.com/mattieb/download-windows-esd/commit/f744452d082bf6539caa8202cd16b016ff390435.

**Comment by elijahgagne** [comment]
Running it from my home network vs corporate network makes no difference. However, the version of `curl` seems to

This fails on my Mac

```
[d92495j:~] % curl -V
curl 8.6.0 (x86_64-apple-darwin23.0) libcurl/8.6.0 (SecureTransport) LibreSSL/3.3.6 zlib/1.2.12 nghttp2/1.61.0
Release-Date: 2024-01-31
Protocols: dict file ftp ftps gopher gophers http https imap imaps ipfs ipns ldap ldaps mqtt pop3 pop3s rtsp smb smbs smtp smtps telnet tftp
Features: alt-svc AsynchDNS GSS-API HSTS HTTP2 HTTPS-proxy IPv6 Kerberos Largefile libz MultiSSL NTLM NTLM_WB SPNEGO SSL threadsafe UnixSockets
```

I used the following commands to instantiate a container and test it

```sh
docker run --rm -ti ubuntu bash
apt -qq update
apt -qq install -y curl
curl -#OR -C - "http://dl.delivery.mp.microsoft.com/filestreamingservice/files/3586ef88-7144-42e3-ba81-bfa35cd48762/22631.2861.231204-0538.23H2_NI_RELEASE_SVC_REFRESH_CLIENTBUSINESS_VOL_A64FRE_en-us.esd"
```

That works fine for me with version

```
root@4996d8100b20:/# curl -V
curl 8.5.0 (aarch64-unknown-linux-gnu) libcurl/8.5.0 OpenSSL/3.0.13 zlib/1.3 brotli/1.1.0 zstd/1.5.5 libidn2/2.3.7 libpsl/0.21.2 (+libidn2/2.3.7) libssh/0.10.6/openssl/zlib nghttp2/1.59.0 librtmp/2.3 OpenLDAP/2.6.7
Release-Date: 2023-12-06, security patched: 8.5.0-2ubuntu10.1
Protocols: dict file ftp ftps gopher gophers http https imap imaps ldap ldaps mqtt pop3 pop3s rtmp rtsp scp sftp smb smbs smtp smtps telnet tftp
Features: alt-svc AsynchDNS brotli GSS-API HSTS HTTP2 HTTPS-proxy IDN IPv6 Kerberos Largefile libz NTLM PSL SPNEGO SSL threadsafe TLS-SRP UnixSockets zstd
```

I haven't been able to find any relevant switches to `curl` to fix things natively on my Mac. 

Oh well, I doubt I'll spend much more time looking into this. Thanks!

**Comment by mattieb** [comment]
Does look like you have the exact same macOS curl I do. If you (or anyone else) lands on any inspiration, happy to revisit this! Thanks for digging.
